lib/archethic/utils/hydrating_cache/cache_entry.ex

```diff
@@ -0,0 +1,229 @@
+  - Run the hydrating function associated with this key
+  - Manage various timers
+  """
+  alias Archethic.Utils.HydratingCache.CacheEntry
+  @behaviour :gen_statem
+
+  use Task
+  require Logger
+
+  def start_link([fun, key, ttl, refresh_interval]) do
+    :gen_statem.start_link(__MODULE__, [fun, key, ttl, refresh_interval], [])
+  end
+
+  @impl :gen_statem
+  def init([fun, key, ttl, refresh_interval]) do
+    timer = :timer.send_interval(refresh_interval, self(), :hydrate)
+
+    ## Hydrate the value
+    {:ok, :running,
+     %CacheEntry.StateData{
+       timer_func: timer,
+       hydrating_func: fun,
+       key: key,
+       ttl: ttl,
+       refresh_interval: refresh_interval
+     }}
+  end
+
+  @impl :gen_statem
+  def callback_mode, do: [:handle_event_function, :state_enter]
+
+  @impl :gen_statem
+  def handle_event({:call, from}, {:get, _from}, :idle, data) do
+    IO.puts("==> Get in idle state")
+    ## Value is requested while fsm is iddle, return the value
+    {:next_state, :idle, data, [{:reply, from, data.value}]}
+  end
+
+  def handle_event(:cast, {:get, from}, :idle, data) do
+    ## Value is requested while fsm is iddle, return the value
+    send(from, {:ok, data.value})
+    {:next_state, :idle, data}
+  end
+
+  ## Call for value while hydrating function is running and we have no previous value
+  ## We register the caller to send value later on, and we indicate caller to block
+  def handle_event({:call, _from}, {:get, from}, :running, data)
+      when data.value == :"$$undefined" do
+    previous_getters = data.getters
+
+    {:keep_state, %CacheEntry.StateData{data | getters: previous_getters ++ [from]},
+     [{:reply, from, {:ok, :answer_delayed}}]}
+  end
+
+  ## Call for value while hydrating function is running and we have a previous value
+  ## We return the value to caller
+  def handle_event({:call, from}, {:get, _from}, :running, data) do
+    {:next_state, :running, data, [{:reply, from, data.value}]}
+  end
+
+  ## Getting value when a function is running and no previous value is available
+  ## Register this getter to send value later on
+  def handle_event(:cast, {:get, from}, :running, data) when data.value == :"$$undefined" do
+    previous_getters = data.getters
+    {:next_state, :running, %CacheEntry.StateData{data | getters: previous_getters ++ [from]}}
+  end
+
+  def handle_event(:cast, {:get, from}, :running, data) do
+    ## Getting value while function is running but previous value is available
+    ## Return vurrent value
+    send(from, {:ok, data.value})
+    {:next_state, :running, data}
+  end
+
+  def handle_event({:call, from}, {:register, fun, key, ttl, refresh_interval}, :running, data) do
+    ## Registering a new hydrating function while previous one is running
+
+    ## We stop the hydrating task if it is already running
+    case data.running_func_task do
+      pid when is_pid(pid) -> Process.exit(pid, :brutal_kill)
```

Comment:
it's probably better to use another signal than brutal kill, to be able to trap the exit and throw some logs